File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example/h10-w/motion/linear_move.cpp
// linear_move.cpp

#include <iostream>
#include <iomanip>
#include "h10-w/h10w_motion.h"
#include "common/common_def.h"
#include "h10-w/h10w_status.h"
#include "h10-w/h10w_system.h"
#include <algorithm>

using namespace humanoid;
using namespace humanoid::h10w;

/**
 * 测试LinearMove，注意：
 * 1、避免运动时发生危险，需要保证机器人四周空旷
 * 2、本测试为单臂运动，运行前control_policy设置为humanoid::ControlPolicy::SINGLE_ARM
 */

int main() {
  std::cout << "=== Linear Move Demo ===" << std::endl;

  H10wMotion motion;
  H10wStatus status;
  H10wSystem system;
  MoveMessageParams moveMsg;
  int result = status.getMoveMessage(moveMsg);
  if (result != 0) {
    std::cerr << "Get move message failed." << std::endl;
    return -1;
  }
  bool leftArmNearZero =
      std::all_of(moveMsg.position.begin(), moveMsg.position.begin() + 7,
                  [](double value) { return value - 0.0 < 0.001; });
  if (!leftArmNearZero) {
    std::cerr << "To ensure safety, please zero the left arm." << std::endl;
    return -1;
  }
  result = system.powerControl(PowerState::POWER_ON);
  if (result != 0) {
    std::cerr << "上电失败" << std::endl;
    return -1;
  }
  result = system.clearError();
  if (result != 0) {
    std::cerr << "清错失败" << std::endl;
    return -1;
  }
  result = system.upperEnableControl(humanoid::EnableState::ENABLE_ON);
  if (result != 0) {
    std::cerr << "上半身使能失败" << std::endl;
    return -1;
  }

  // move to start position
  // ->tcp:{1.225196,0.387371,1.157848,0.683798,2.018945,0.074912}
  std::vector<JointTarget> leftStartPos{
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT1), -0.2},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT2), 0.2},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT3), 0.2},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT4), 0.9},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT5), 0.8},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT6), 0.5},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT7), 0.5},
      {static_cast<uint32_t>(JointIndexValue::ELEVATOR_MOTOR), 0.3}};
  result = motion.multiJointsMove(leftStartPos, 0.1, 0.1);
  if (result != 0) {
    std::cout << "Move to start position failed." << std::endl;
    return -1;
  }
  result = motion.waitMove();

  if (result != 0) {
    std::cout << "Wait move to start position failed." << std::endl;
    return -1;
  }

  // x:1.225196->1.025196
  std::vector<LinearTarget> linearTargets = {
      {CartIndex::LEFT_ARM,  // 左臂
       {1.025196, 0.387371, 1.157848, 0.683798, 2.018945,
        0.074912},  // 目标位置
       0.1,         // 速度 10%
       0.1}         // 加速度 10%
  };

  if (motion.linearMove(linearTargets) == 0) {
    std::cout << "Start linear move success." << std::endl;
  } else {
    std::cout << "Start linear move failed." << std::endl;
    return -1;
  }

  result = motion.waitMove();

  if (result != 0) {
    std::cout << "Wait linear move failed." << std::endl;
    return -1;
  }

  std::vector<JointTarget> leftZeroPos{
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT1), 0},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT2), 0},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT3), 0},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT4), 0},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT5), 0},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT6), 0},
      {static_cast<uint32_t>(JointIndexValue::L_ARM_JOINT7), 0}};
  result = motion.multiJointsMove(leftZeroPos, 0.1, 0.1);
  if (result != 0) {
    std::cout << "Start move to zero failed." << std::endl;
  }
  result = motion.waitMove();

  if (result != 0) {
    std::cout << "Wait move to zero failed." << std::endl;
  }

  return 0;
}

File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example_py/h10-w/motion/linear_move.py
import humanoid_sdk_py
from humanoid_sdk_py import h10w
import time
h10w_motion = h10w.H10wMotion()
h10w_status = h10w.H10wStatus()
h10w_system = h10w.H10wSystem()
time.sleep(1)

print("=== 线性移动示例 ===")

ret, move_msg = h10w_status.getMoveMessage()
if ret != 0 or move_msg is None:
    print("获取运动信息失败，任务终止")
    exit(-1)

left_arm_near_zero = all(abs(value - 0.0) < 0.001 for value in move_msg.position[:7])
if not left_arm_near_zero:
    print("为保证安全，请先将左臂回零")
    exit(-1)

ret = h10w_system.powerControl(humanoid_sdk_py.PowerState.POWER_ON)
if ret != 0:
    print("上电失败")
    exit(-1)

ret = h10w_system.clearError()
if ret != 0:
    print("清除错误失败")
    exit(-1)

ret = h10w_system.upperEnableControl(humanoid_sdk_py.EnableState.ENABLE_ON)
if ret != 0:
    print("上半身使能失败")
    exit(-1)

# move to start position
left_start_pose = [
    (h10w.L_ARM_JOINT1, -0.2),
    (h10w.L_ARM_JOINT2, 0.2),
    (h10w.L_ARM_JOINT3, 0.2),
    (h10w.L_ARM_JOINT4, 0.9),
    (h10w.L_ARM_JOINT5, 0.8),
    (h10w.L_ARM_JOINT6, 0.5),
    (h10w.L_ARM_JOINT7, 0.5),
    (h10w.ELEVATOR_MOTOR, 0.3),
]
ret = h10w_motion.multiJointsMove(left_start_pose, 0.1, 0.1)
if ret != 0:
    print("运动指令发送失败")
    exit(-1)
ret = h10w_motion.waitMove()
if ret != 0:
    print("运动失败")
    exit(-1)

# move to target pose
target = h10w.LinearTarget()
target.type = h10w.CartIndex.LEFT_ARM
target.pose = [1.025196, 0.387371, 1.157848, 0.683798, 2.018945, 0.074912]
target.velocityPercent = 0.1  # 10% of max velocity
target.accelerationPercent = 0.1  # 10% of max acceleration
linear_targets = [target]
ret = h10w_motion.linearMove(linear_targets)
if ret != 0:
    print("运动指令发送失败")
else:
    print("运动指令发送成功")

ret = h10w_motion.waitMove()
if ret != 0:
    print("运动失败")
    exit(-1)

left_arm_homing = [
    (h10w.L_ARM_JOINT1, 0.0),
    (h10w.L_ARM_JOINT2, 0.0),
    (h10w.L_ARM_JOINT3, 0.0),
    (h10w.L_ARM_JOINT4, 0.0),
    (h10w.L_ARM_JOINT5, 0.0),
    (h10w.L_ARM_JOINT6, 0.0),
    (h10w.L_ARM_JOINT7, 0.0),
]
ret = h10w_motion.multiJointsMove(left_arm_homing, 0.1, 0.1)
if ret != 0:
    print("运动指令发送失败")
    exit(-1)

ret = h10w_motion.waitMove()
if ret != 0:
    print("运动失败")
    exit(-1)
